TechnicalDocuments/Proposals/StyleSheets.md

```diff
@@ -0,0 +1,119 @@
+
+    In the case of a reusable view/cell, the expectation is that the view would store the inverse/snapshot & apply it as part of the cleanup e.g. in `prepareForReuse()`.
+    
+    This eliminates the state indeterminism of closure-based styling approaches, while not requiring any default value to be declared upfront like how our `StyleSheet` framework currently does. 
+
+4. Can be used for any view and any arbitrary subclass without type declaration.
+
+### Implementation Plan
+
+1. As soon as the new `StyleSheet` type is available in Bento, new components shall start using it.
+
+2. Extensions to either `UIView` or `Base*View` shall be made to provide a convenience which stores & applies style sheets on one's behalf. For example:
+
+   ```swift
+   view.backgroundColor = .green
+   view.clipsToBound = false
+
+    // NOTE: `apply` automatically keeps the inverse.
+   view.apply(
+       StyleSheet()
+            .setting(\.backgroundColor, .red)
+            .setting(\.clipsToBound, true)
+    )
+   expect(view.backgroundColor) == .red
+   expect(view.clipsToBound) == true
+
+    // NOTE: The stored inverse is applied before the new style sheet is applied.
+    //       So `clipsToBound` would be reverted to the original value.
+   view.apply(
+       StyleSheet()
+            .setting(\.backgroundColor, .blue)
+    )
+
+   expect(view.backgroundColor) == .blue
+   expect(view.clipsToBound) == false
+   ```
+
+3. BentoKit components shall be gradually migrated as soon as possible, potentially in batches to limit source breakage. Existing components may continue to use their existing style sheet types until author wishes to migrate.
+
+### Potential Issues
+#### Support only mutable properties.
+Not all stying parameters are exposed as instance properties.
+
+`UIButton` is the most apparent example. We could mitigate this by providing extensions with computed properties of `[UIControl.State: U]`.
+
+#### Might require use of protocols to limit exposed surface.
+If there are properties, inherited from a parent view class, that should be hidden on the component root view, one might need to declare a protocol with all whitelisted properties.
+
+```swift
+protocol SpecialButtonProtocol {
+    var layoutMargins: UIEdgeInsets { get set }
+}
+
+extension UIButton: SpecialButtonProtocol {}
+
+// EXAMPLE: We hide all properties except for `layoutMargins`.
+let styleSheet: StyleSheet<SpecialButtonProtocol>
+
+var button = UIButton()
+styleSheet.apply(to: &button)
+```
+
+## Impact on existing APIs
+The value type `StyleSheet` is an additive change alone.
+
+However, assuming we subsequently replace all existing BentoKit component style sheets, this shall be a massive source breaking change.
+
+### Potential Migration Path
+While script-based and staged migration is a viable solution, additional effort on top of snapshot testing might be required to verify that the removal of style sheet defaults does not lead to changes in visual appearance. 
```

Comment:
We could suggest to use the `@available(*, renamed: ...)` attribute to facilitate the migration via fix-its?

Reply:
It could be a possible to have both old stylesheets to co-exist, and then Bento components could do that. But I haven’t explored that path yet.